plugins/inputs/solr/solr.go

```
@@ -0,0 +1,395 @@
+			"optimizes":                   metrics.Stats.Optimizes,
+			"rollbacks":                   metrics.Stats.Rollbacks,
+			"soft_autocommits":            metrics.Stats.SoftAutocommits,
+		}
+		acc.AddFields(
+			"solr_mbean_metrics",
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"type":    strings.ToLower(category),
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func gatherCacheMetrics(mbeansJson json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]Cache
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJson, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		coreFields := map[string]interface{}{
+			"class_name":           metrics.Class,
+			"cumulative_evictions": metrics.Stats.CumulativeEvictions,
+			"cumulative_hitratio":  metrics.Stats.CumulativeHitratio,
+			"cumulative_hits":      metrics.Stats.CumulativeHits,
+			"cumulative_inserts":   metrics.Stats.CumulativeInserts,
+			"cumulative_lookups":   metrics.Stats.CumulativeLookups,
+			"evictions":            metrics.Stats.Evictions,
+			"hitratio":             metrics.Stats.Hitratio,
+			"hits":                 metrics.Stats.Hits,
+			"inserts":              metrics.Stats.Inserts,
+			"lookups":              metrics.Stats.Lookups,
+			"size":                 metrics.Stats.Size,
+			"warmup_time":          metrics.Stats.WarmupTime,
+		}
+		acc.AddFields(
+			"solr_mbean_metrics",
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"type":    strings.ToLower(category),
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func (s *Solr) gatherCoreStats(url string, core string, acc telegraf.Accumulator) error {
+
+	for _, category := range []string{"CORE", "QUERYHANDLER", "UPDATEHANDLER", "CACHE"} {
+		metrics := &Metrics{}
+		if err := s.gatherData(fmt.Sprintf("%s&cat=%s", url, category), metrics); err != nil {
+			return err
```

Can we use the `Accumulator.AddError()` interface instead of returning, that way even if one "category" fails, the others might still work.